guava/src/com/google/common/eventbus/SubscriberRegistry.java

```diff
@@ -175,27 +175,35 @@ void unregister(Object listener) {
 
   private static ImmutableList<Method> getAnnotatedMethodsNotCached(Class<?> clazz) {
     Set<? extends Class<?>> supertypes = TypeToken.of(clazz).getTypes().rawTypes();
-    Map<MethodIdentifier, Method> identifiers = Maps.newHashMap();
-    for (Class<?> supertype : supertypes) {
-      for (Method method : supertype.getDeclaredMethods()) {
-        if (method.isAnnotationPresent(Subscribe.class) && !method.isSynthetic()) {
-          // TODO(cgdecker): Should check for a generic parameter type and error out
-          Class<?>[] parameterTypes = method.getParameterTypes();
-          checkArgument(
-              parameterTypes.length == 1,
-              "Method %s has @Subscribe annotation but has %s parameters."
-                  + "Subscriber methods must have exactly 1 parameter.",
-              method,
-              parameterTypes.length);
-
-          MethodIdentifier ident = new MethodIdentifier(method);
-          if (!identifiers.containsKey(ident)) {
-            identifiers.put(ident, method);
+    Set<MethodIdentifier> identifiers = new HashSet<>();
+    return supertypes
+      .stream()
+      .filter(cls -> !Object.class.equals(cls))
+      .flatMap(supertype -> {
+        Method[] methods;
+        try {
+          methods = supertype.getDeclaredMethods();
+        } catch (SecurityException e) {
+          methods = supertype.getMethods();
```

Comment:
using `getMethods` will call anothor bug I think.
`getMehods` will get all the methods from subclass, include its superclass.  so when code like that:
```
class ListenerA {
  @Subscribe void listenA(Event event) {}
} 

class ListenerB extends ListenerA {
  @Subscribe void listenB(Event event) {}
}
```

if we use `ListenerB.class.getMethods`，we will get `ListenerA.listenA` as well. it will be a bug, in this code, we will get 2 method obj of `ListenersA.listenA`, one froms ListenerA it self, another froms ListenerB

Reply:
Please explain why that will be a bug?

Reply:
sorry, I ignored the `MethodIdentifier filter` you writed at last, I had make a test for my comment!  : )
```
identifiers = HashSet();
.filter(method -> identifiers.add(new MethodIdentifier(method)))
```
and here is my test code.
```java
  public void testSuperAnsSubSubscribe() {
    SubClass subClass = new SubClass();
    registry.register(subClass);
    assertEquals(2, Iterators.size(registry.getSubscribers(Integer.valueOf(1))));
  }

  private static class SuperClass {
    @Subscribe
    public void superHandle(Integer i) {}
  }

  private static class SubClass extends SuperClass {
    @Subscribe
    public void subHandle(Integer i) {}
  }
```